### DISCUSSION TITLE
How do I set the PWM frequency?

### DISCUSSION THREAD

**Discussion by ptunstall** [Moddable-OpenSource/moddable#1327] [Q&A] [answered]
Hi how do I manually define my PWM frequency for both servo and PWM?  i see in the pwm.c file there's a "freq_hz" being set via some config variable.  Is it possible to define this in the manifest and/or function?

**Comment by phoddie** [comment, marked as answer]
Use the `hz` property as described [here](https://419.ecma-international.org/#-10-io-classes-pulse-width-modulation):

```js
new device.io.PWM({
   pin: 12,
   hz: 5000
});
```

**Reply by ptunstall** [reply]
That appears to work! Is it possible to do this with Servo?

**Reply by phoddie** [reply]
> That appears to work! 

Great.

> Is it possible to do this with Servo?

Doesn't [look like it](https://github.com/Moddable-OpenSource/moddable/blob/fc5c2b41c0e2df4f9235f11aa1962886a603d4dd/modules/pins/servo/esp32/modServo.c#L60C9-L60C11).  Should be pretty straightforward to add though.

**Reply by ptunstall** [reply]
Actually I take this back I don't see the HZ property doing anything.  Possibly my interpretation is incorrect?

```
import Servo from "pins/servo";
import PWM from "pins/pwm";
import Analog from "pins/analog";
import Timer from "timer";

//let servo = new Servo({pin: 5, min: 0, max: 2000});

let pwm = new PWM({ pin: 5, hz: 50, min: 0, max: 2000 });


let angle = 0;
let step = 1;

let R1 = 3950;
let c1 = 1.009249522e-03;
let c2 = 2.378405444e-04;
let c3 = 2.019202697e-07;
let T = 0;

let initilize = 0;

//Timer.repeat(id => {
//	let value1 = Analog.read(0);
//	let R2 = R1 * (1023/(value1-1));
//	let logR2 = Math.log(R2);
//	T = (1/(c1 + c2*logR2 + c3*logR2*logR2*logR2));
//	T -= 273.15;
//	T = (T*9)/5+32;
//	trace(`value1: ${value1}\n`);
//	trace(`Temp F: ${T}\n`);
//	if (initilize == 1)
//		servo.write(T);
//	let value2 = Analog.read(2);
//	trace("value1: " + value1 + " value2: " + value2 + "\n");
//}, 50);

/* Timer.repeat(() => {
	angle += step;
	if (angle > 180 || angle < 0) 
		step *= -1;
	servo.write(angle);
	trace(`value: ${angle}\n`);
}, 50); */

Timer.repeat(() => {
	angle += step;
	if (angle >= 511 || angle <= 0)
		step *= -1;
	pwm.write(angle)
	trace(`value: ${angle}\n`);
}, 20);
```
We're trying to drive PWM controlled fans for cooling a device.  These appear to require a 25000hz frequency for the PWM signal. 

I dropped the HZ to 50 see if I could see it flickering to the naked eye.

**Reply by phoddie** [reply]
You are using a different PWM implementation than the one my example references. You are using the older "pins" version; the example I gave is from the newer ECMA-419 standard APIs. Try the [LED PWM example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/pwm/led/main.js) to get started with ECMA-419 PWM. There's also a [servo example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/pwm/servo-sweep/main.js) using ECMA-419 PWM.

**Reply by ptunstall** [reply]
YES!!!! Success! You are amazing!! Super appreciate the continued help and patience!
